common/unsafe/src/main/java/org/apache/spark/unsafe/memory/UnsafeMemoryAllocator.java

```diff
@@ -19,28 +19,48 @@
 
 import org.apache.spark.unsafe.Platform;
 
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.nio.ByteBuffer;
+
+import sun.nio.ch.DirectBuffer;
+
 /**
  * A simple {@link MemoryAllocator} that uses {@code Unsafe} to allocate off-heap memory.
  */
 public class UnsafeMemoryAllocator implements MemoryAllocator {
 
   @Override
-  public MemoryBlock allocate(long size) throws OutOfMemoryError {
-    long address = Platform.allocateMemory(size);
-    MemoryBlock memory = new MemoryBlock(null, address, size);
-    if (MemoryAllocator.MEMORY_DEBUG_FILL_ENABLED) {
-      memory.fill(MemoryAllocator.MEMORY_DEBUG_FILL_CLEAN_VALUE);
+  public OffHeapMemoryBlock allocate(long size) throws OutOfMemoryError {
+    Object buffer = ByteBuffer.allocateDirect((int)size);
```

Comment:
what's the different between `Platform.allocateMemory` and `ByteBuffer.allocateDirect`?

Reply:
`Platform.allocateMemory` return a memory address of the region in off-heap. This region will be freed only when we call `Platform.freeMemory`.  
`ByteBuffer.allocateDirect` return a Java object that wraps a memory region in off-heap. This region will be automatically freed when the Java object is collected by GC.

Reply:
I knew that to use `Platform.allocateMemory` is [current design](http://downloads.typesafe.com/website/presentations/ScalaDaysSF2015/T4_Xin_Performance_Optimization.pdf#page=26). I will revert this change.

If we want to use `ByteBuffer.allocateDirect`, it would be good to submit a separate PR.

Reply:
To address issue for releasing a memory, [here](https://hbase.apache.org/0.94/apidocs/src-html/org/apache/hadoop/hbase/util/DirectMemoryUtils.html#line.80) is one solution

Reply:
I just wanna understand why we would like to pick `ByteBuffer.allocateDirect` with a hacky free method. Does `ByteBuffer.allocateDirect` have better performance? or you just wanna change the life cycle of `MemoryBlock` and ask GC to control it?

Reply:
This is easy to manage life cycle of native memory without calling `free`. This code was derived from #11494.